### PR TITLE
Add jwt header to websocket handshake.

### PR DESCRIPTION
**Pull request by menakaj**
### Proposed changes in this pull request
- Generate the jwt token in WS key validation.
- Attach the jwt token as a custom header.
- Add test cases.

### When should this PR be merged
Please merge the PR after merging wso2/carbon-mediation/pull/1068

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java [hunk 1]
```diff
@@ -158,6 +160,11 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                 } else {
                     req.setUri(uri); // Setting endpoint appended uri
                 }
+
+                if (!Strings.isNullOrEmpty(token)) {
+                    ((FullHttpRequest) msg).headers().set(APIMgtGatewayConstants.WS_JWT_TOKEN_HEADER, token);
+
```
Comment (Megala21) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java]: Please remove extra line

Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java]: Fixed in https://github.com/wso2/carbon-apimgt/pull/5519/commits/55935bebfa203b2967f9ba165a730014a431e2d2

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyValidationService.java [hunk 2]
```diff
@@ -316,6 +317,18 @@ public APIKeyValidationInfoDTO validateKeyforHandshake(String context, String ve
             if (defaultVersionInvoked) {
                 info.setApiName(info.getApiName() + "*" + version);
             }
+
+            if (APIKeyMgtDataHolder.isJwtGenerationEnabled() &&
+                    validationContext.getValidationInfoDTO().getEndUserName() != null
+                    && !validationContext.isCacheHit()) {
+                Application application = APIUtil.getApplicationByClientId(validationContext.getValidationInfoDTO()
+                        .getConsumerKey());
+
```
Comment (isurulucky) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyValidationService.java]: remove the extra line

Comment (isurulucky) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyValidationService.java]: remove this extra line.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/APIMgtGatewayConstants.java [hunk 3]
```diff
@@ -90,5 +90,9 @@
     public static final String XSD_URL = "xsdURL";
     public static final String UTF8 = "UTF-8";
 
+    /**
+     * Web socket header for jwt assertion.
+     * */
+    public static final String WS_JWT_TOKEN_HEADER = "websocket.custom.header.X-JWT-Assertion";
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/APIMgtGatewayConstants.java]: Why can't we use the same standard header name "X-JWT-Assertion"?

Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/APIMgtGatewayConstants.java]: Hi Fazlan,

The headers in the client handshake are not sent to the backend directly. In the WebSocketTransportSender, the headers are added to the axis2 message context as properties. There we cannot filter the headers that should be sent to the backend. So, we prefix the actual header that should be sent to the backend and add it as a header.

In the TransportSender, the headers are filtered by the prefix, extract the actual header and attach it to the handshake request. 

So in this scenario, the actual header that is sent to the backend is X-JWT-Assertion.

**Pull request merged**